### PR TITLE
:sparkles: add custom variables and code snippets 

### LINKED ISSUE #31: We need to figure out what File URI / URL is supposed to based on
We need to determine the root here and standardize the output

### LINKED ISSUE #117: Rule Authors Define Custom Extras
Right now, we have extras that can be used in the message for users, but a user may want to add their own.

this can be useful, in that a rule author now does not have to change the provider of the capability to eloquently write complex rules. We see examples of this use case in the eapX/websphere rules in windups rulesets.

### PR DESCRIPTION
**Pull request by shawn-hurley**
fixes #31 
fixes #117

### CODE REVIEW

File: hubapi/violations.go [hunk 1]
```diff
@@ -60,9 +62,10 @@ type Violation struct {
...
 	// URI defines location in the codebase where violation is found
-	URI string `yaml:"uri"`
+	URI uri.URI `yaml:"uri"`
 	// Message text description about the incident
-	Message string `yaml:"message"`
+	Message  string `yaml:"message"`
+	CodeSnip string `yaml:"codeSnip,omitempty"`
 	// Extras reserved for additional data
 	//Extras json.RawMessage
 	Extras map[string]interface{}
```
Comment (shawn-hurley) [inline hubapi/violations.go:71]: Change this name to Variables too

**Comment by shawn-hurley** [comment]
Still a Draft PR as I need to work on the custom variables, but the current implementation gives code snips.

File: engine/conditions.go [hunk 2]
```diff
@@ -35,10 +37,36 @@ type ConditionEntry struct {
 }
 
 type IncidentContext struct {
-	FileURI string                 `yaml:"fileURI"`
-	Effort  *int                   `yaml:"effort"`
-	Extras  map[string]interface{} `yaml:"extras"`
-	Links   []hubapi.Link          `yaml:"externalLink"`
+	FileURI      uri.URI                `yaml:"fileURI"`
```
Comment (fabianvf) [inline engine/conditions.go:40]: nice

**Review by fabianvf — APPROVED** [review]
looks good

File: demo-output.yaml [hunk 3]
```diff
@@ -136,96 +136,66 @@
...
-      - uri: provider/lib/lib.go
+      - uri: file:///analyzer-lsp/provider/lib/lib.go
         message: all go files
-        extras: {}
-      - uri: provider/provider.go
+      - uri: file:///analyzer-lsp/provider/provider.go
         message: all go files
-        extras: {}
-      - uri: provider/provider_test.go
+      - uri: file:///analyzer-lsp/provider/provider_test.go
```
[the remark this answers is not in the document]
Comment (shawn-hurley) [reply demo-output.yaml:195]: I have gone back and forth and I think we may need to different things for different reasons

I think provider to engine has to be abs path on filesystem. And then what we tell the hub can be relative to base of project. I wonder if we can just use the provider settings strip off the dir and say that becomes the base?

**Pull request merged**